rfd/0163-vnet.md

```diff
@@ -0,0 +1,416 @@
+
+TODO: Dedicated button in top left with status indicator.
+
+#### Lifecycle
+
+The first version of VNet integration in Connect is going to behave similarly to Connect My
+Computer. After the user starts VNet and until they explicitly stop it, VNet will start
+automatically when launching the app. It will prompt for password for admin privileges through
+osascript on each start. This requires the given system user to be an admin, disqualifying cases
+where the user is just a standard macOS account.
+
+Unfortunately, [Touch ID prompts are reserved for Apple
+clients](https://forums.developer.apple.com/forums/thread/701118?answerId=705661022#705661022). We
+assume that typically the user is going to start VNet and then just leave it running, somewhat
+negating the downside of having to enter credentials on each start.
+
+##### Considered alternatives
+
+Tying the lifecycle of VNet to the lifecycle of Connect gets us past the problem of needing root
+privileges to create a TUN device and lets us focus on validating VNet as a tool. However, there is
+a couple of different approaches we could take which offer better UX at the cost of development
+time.
+
+###### Launch VNet on system start
+
+Doing so would require some kind of persistent UI that is also launched on system start, think an
+app in the Menu Bar. The UI would need to enable the user to manage the state of VNet and provide a
+way to refresh expired certs. Connect could support headless mode where it does not open the main
+window automatically and instead launches a separate frontend app which handles relogin.
+
+###### Ask for admin privileges just once
+
+The recommended way for an unprivileged app to use privileged APIs is to use
+[`SMAppService`](https://developer.apple.com/documentation/servicemanagement/smappservice).
+`SMAppService` allows the unprivileged app to add a privileged launch daemon to launchd. That daemon can be
+launched on demand from the unprivileged, do its work and then shut down, without having to enter
+credentials each time. The daemon lives in the app bundle and is automatically cleaned up when the
+user removes `Teleport Connect.app`.
+
+With `SMAppService`, there are two main problems to figure out:
+
+* How to call this Swift API in the first place.
+  * Do we need to build our own simple app just to register a daemon for the main app?
+* How to securely communicate with the daemon.
+  * Communication is typically done through XPC, as shown in [the example
+    app](https://developer.apple.com/documentation/servicemanagement/updating-your-app-package-installer-to-use-the-new-service-management-api).
+    How viable it is to use it from Go?
+  * Secure communication can be achieved through [environment
+    constraints](https://developer.apple.com/videos/play/wwdc2023/10266/) by enforcing that only
+    programs signed by us can interact with the daemon.
+  * Is gRPC a viable alternative here?
```

Comment:
During the mobile app hackathon, we were able to create a mTLS gRPC client in Swift (using https://github.com/apple/swift-nio-ssl) - I imagine creating a server would be similar.

Reply:
The daemon most probably could be just a Go program that sets up a gRPC server. The problem here is not as much with setting up the server itself, but with making sure that no other program on the machine can talk to the daemon that's running as root.

With XPC, [you can check how the other process has been codesigned](https://github.com/aronskaya/smjobbless/blob/master/SecuringXPCConnection.md) to make sure it was signed by the same Apple ID cert for example.

With gRPC, we'd need to find another way. Nothing comes to my mind off the top of my head and the first version is not going to use a daemon anyway, hence why I didn't include many details here. I should probably rephrase this section to make it more clear.